File: core/src/main/scala/tripaint/image/format/SimpleStorageFormat.scala
package tripaint.image.format

import tripaint.coords.{StorageCoords, TriangleCoords}

object SimpleStorageFormat extends StorageFormat {
  override def transform(coords: TriangleCoords): StorageCoords =
    val x = coords.x
    val y = coords.y
    if y < x
    then StorageCoords(y, y + y - x)
    else StorageCoords(x, y)

  override def reverse(coords: StorageCoords): TriangleCoords =
    val x = coords.x
    val y = coords.y
    if y < x
    then TriangleCoords(x + x - y, x)
    else TriangleCoords(x, y)
}

File: core/src/main/kotlin/tripaint/image/format/SimpleStorageFormat.kt
package tripaint.image.format

import tripaint.coords.StorageCoords
import tripaint.coords.TriangleCoords

object SimpleStorageFormat : StorageFormat {
    override fun supportsImageSize(size: Int): Boolean {
        return true
    }

    override fun transform(coords: TriangleCoords): StorageCoords {
        val x = coords.x
        val y = coords.y
        return if (y < x) {
            StorageCoords.from(y, y + y - x)
        } else {
            StorageCoords.from(x, y)
        }
    }

    override fun reverse(coords: StorageCoords): TriangleCoords {
        val x = coords.x
        val y = coords.y
        return if (y < x) {
            TriangleCoords.from(x + x - y, x)
        } else {
            TriangleCoords.from(x, y)
        }
    }
}
